### PR TITLE
`ls`: add support for showing SELinux context (--context/-Z)

### PR DESCRIPTION
**Pull request by jhscheer**
This adds SELinux security context (--context/-Z) support to `uu_ls`.

Since the SELinux specific parts are  gated behind `--featues feat_selinux`, the tests don't run in the CICD and I appreciate a volunteer to run:
`cargo test test_ls_context --features feat_selinux`
on an SELinux enabled Linux distribution with libselinux headers installed (e.g. on fedora dnf install libselinux-devel).

(With this the "gnu/tests/id/no-context.sh" test now passes because it depends on `uu_ls -Z` being implemented.)

### CODE REVIEW

File: src/uu/ls/src/ls.rs [hunk 1]
```diff
@@ -611,6 +639,10 @@ pub fn uumain(args: impl uucore::Args) -> UResult<()> {
         .map(|v| v.map(Path::new).collect())
         .unwrap_or_else(|| vec![Path::new(".")]);
 
+    if config.context && !config.selinux_supported {
+        show_warning!("--context (-Z) works only on an SELinux-enabled kernel",);
+    }
+
```
Comment (blyxxyz) [inline src/uu/ls/src/ls.rs:645]: GNU ls doesn't seem to do this, so perhaps there should only be a warning if the flag is used on a `uu_ls` build without the feature enabled?

Comment (jhscheer) [reply src/uu/ls/src/ls.rs:645]: Yes, good point. I'll do that.

**Comment by sylvestre** [comment]
@koutheir FYI ;)

**Comment by koutheir** [comment]
Thank you @jhscheer for the efforts!

I have the impression that some weird behavior happens when an SELinux context is not valid UTF-8. Sometimes it will be a warning, and other times output with UTF-8 replacement characters. Is is possible to drop the assumption of UTF-8 on contexts?

**Comment by jhscheer** [comment]
Oh bummer. I assumed  the SELinux Context strings are UTF8.
I'll try to fix that.

**Comment by koutheir** [comment]
Because of the (maybe intended) lack of precision in the documentation of `libselinux`, the `selinux` crate assumes that SELinux contexts are represented as:
- Either a sequence of bytes, when the `libselinux` API providing the context returns the byte pointer and its size.
- Or a null-terminated C string, when the `libselinux` API providing the context returns only a pointer.

This is why a [`selinux::SecurityContext`](https://docs.rs/selinux/0.2.3/selinux/struct.SecurityContext.html) instance returns either a byte slice representation of itself (See [`selinux::SecurityContext::as_bytes()`](https://docs.rs/selinux/0.2.3/selinux/struct.SecurityContext.html#method.as_bytes)) or a `CStr` representation of itself (See [`selinux::SecurityContext::to_c_string()`](https://docs.rs/selinux/0.2.3/selinux/struct.SecurityContext.html#method.to_c_string)).

Assuming UTF-8 encoding might be fine for an initial simple but limited implementation of the `--context` flag, but a full implementation should not assume this. Please tell me if you think I missed something in the `libselinux`/SELinux documentation that specifies the format of contexts.

I understand that `ls` is in the business of printing information, so ultimately, an SELinux context must be printable. However, this is unrelated to the context being a valid UTF-8 string, because C's `printf()` doesn't require that. I think that, if we need to print a context that is not UTF-8-encoded, then we should both:
- Use replacement characters (because a person prefers to see some characters of the context, instead of nothing), and
- Issue a warning informing that the printed context cannot be used as is (for instance, by copying it) in other SELinux-related operations that require an input context, because it is only partially printed.

Again, please tell me if I missed something.

**Comment by jhscheer** [comment]
Thanks for the deeper insight.

I implemented what you proposed:
https://github.com/uutils/coreutils/blob/cd2153eac624232ff240b323eb4c04c110308e53/src/uu/ls/src/ls.rs#L1272-L1286

I wasn't able to create a file with a security context which contains invalid utf8.
Do you have an idea how to tinker a test-case for this?

**Comment by blyxxyz** [comment]
> I think that, if we need to print a context that is not UTF-8-encoded, then we should both:
> 
> * Use replacement characters (because a person prefers to see some characters of the context, instead of nothing), and

`gnome-terminal` displays invalid UTF-8 as replacement characters already, as does `xterm`.

If the output is captured by some program then printing the original bytes is better.

If the output is to be shown to humans then it could be conditionally quoted, either by one of `ls`'s existing quoting styles or by [`Quotable::maybe_quote()`](https://uutils.github.io/coreutils-docs/uucore/display/trait.Quotable.html#method.maybe_quote). But that might also quote `?` so that'd have to be handled specially. Maybe an unknown/missing context could be represented by `None` until it's printed?

`ls` currently isn't actually capable of printing non-UTF-8, but there's [`print_verbatim`](https://uutils.github.io/coreutils-docs/uucore/display/fn.print_verbatim.html) for that.

In the long term I think it'd make the most sense to [convert](https://doc.rust-lang.org/std/os/unix/ffi/trait.OsStringExt.html#tymethod.from_vec) the context into an `OsString` and change `ls` to handle those better. There's some discussion of that in #2621.

**Comment by koutheir** [comment]
> I wasn't able to create a file with a security context which contains invalid utf8.
> Do you have an idea how to tinker a test-case for this?

I also could not create one (also for `selinux` unit tests). This is one of the nice things one can have by mocking the `libselinux` API, but that seems too large of a task to do accurately.

**Comment by koutheir** [comment]
> If the output is captured by some program then printing the original bytes is better.

I agree.

> In the long term I think it'd make the most sense to convert the context into an `OsString` and change `ls` to handle those better. There's some discussion of that in #2621.

It is already possible to represent SELinux contexts as an `OsStr`, using the UNIX-specific [`OsStrExt::from_bytes`](https://doc.rust-lang.org/std/os/unix/ffi/trait.OsStrExt.html#tymethod.from_bytes).

File: src/uu/ls/src/ls.rs [hunk 2]
```diff
@@ -1224,12 +1259,53 @@ impl PathData {
             None => OnceCell::new(),
         };
 
+        let substitute_string = "?".to_string();
+        let security_context = if config.context {
+            if config.selinux_supported {
+                #[cfg(feature = "selinux")]
+                {
+                    match selinux::SecurityContext::of_path(&p_buf, must_dereference, false) {
```
Comment (sylvestre) [inline src/uu/ls/src/ls.rs:1267]: what do you think about moving this block into a new function to simplify a bit the code?

Comment (jhscheer) [reply src/uu/ls/src/ls.rs]: Sounds good. Done.

File: src/uu/ls/src/ls.rs [hunk 3]
```diff
@@ -1992,3 +2086,47 @@ fn display_symlink_count(_metadata: &Metadata) -> String {
 fn display_symlink_count(metadata: &Metadata) -> String {
     metadata.nlink().to_string()
 }
+
+// This returns the SELinux security context as UTF8 `String`.
+// In the long term this should be changed to `OsStr`, see discussions at #2621/#2656
+fn get_security_context(config: &Config, p_buf: &PathBuf, must_dereference: bool) -> String {
```
Comment (sylvestre) [inline src/uu/ls/src/ls.rs:2092]: you have some warnings here :)

Comment (sylvestre) [reply src/uu/ls/src/ls.rs:2092]: Feels free to silent the warning if it increases the complexity

Comment (jhscheer) [reply src/uu/ls/src/ls.rs]: done

**Comment by sylvestre** [comment]
thanks for the ping :)

**Pull request merged**